descriptions.go

```diff
@@ -126,11 +131,255 @@ type RequestorInfo struct {
+	skipped := 0
+	for i := len(contents) - 1; i >= 0; i-- {
+		item := contents[i]
+		if item.Credential == nil {
+			skipped++
+			continue
+		}
+		reversed = append(reversed, contents[i])
+		byID[*contents[i].Credential] = contents[i]
+	}
+
+	// Build a map containing per level of the dependency tree the (deduplicated) nodes at that level
+	// (root = level 0, its dependencies = level 1, their dependencies = level 2, etc).
+	deps := credentialDependencies{}
+	bylevel := map[int]map[CredentialTypeIdentifier]struct{}{}
+	for i, item := range reversed {
+		if err := wizardItemVisit(bylevel, i, *item.Credential, conf, deps, credsmap); err != nil {
+			return nil, err
+		}
+	}
+
+	// Scanning horizontally, i.e. per level, we iterate across the tree, putting all
+	// credential types that we come across in the result slice. This starts at the leaf nodes
+	// that have no dependencies, and after that across the intermediate nodes whose dependencies
+	// have been put in the result slice in previous iterations.
+	var result []IssueWizardItem                         // to return
+	resultmap := map[CredentialTypeIdentifier]struct{}{} // to keep track of credentials already put in the result slice
+	for i := len(bylevel) - 1; i >= 0; i-- {
+		for id := range bylevel[i] {
+			if _, ok := resultmap[id]; ok {
+				continue
+			}
+			resultmap[id] = struct{}{}
+			if item, present := byID[id]; present {
+				result = append(result, item)
+			} else {
+				current := id // create copy of loop variable to take address of
+				result = append(result, IssueWizardItem{
+					Type:       IssueWizardItemTypeCredential,
+					Credential: &current,
+				})
+			}
+		}
+	}
+
+	result = append(result, contents[len(contents)-skipped:]...)
+	return result, nil
+}
+
+type credentialDependencies map[CredentialTypeIdentifier][]IssueWizardItem
+
+// wizardItemVisit is a recursive function that populates a map containing per level of a tree the
+// (deduplicated) nodes at that level.
+func wizardItemVisit(
+	bylevel map[int]map[CredentialTypeIdentifier]struct{},
+	level int,
+	id CredentialTypeIdentifier,
+	conf *Configuration,
+	deps credentialDependencies,
+	creds map[CredentialTypeIdentifier]struct{},
+) error {
+	if level == 25 {
+		return errors.New("issue wizard: level limit exceeded, probably due to circular dependency")
+	}
+
+	if bylevel[level] == nil {
+		bylevel[level] = map[CredentialTypeIdentifier]struct{}{}
+	}
+	bylevel[level][id] = struct{}{}
+
+	for _, child := range deps.get(id, conf, creds) {
+		if err := wizardItemVisit(bylevel, level+1, *child.Credential, conf, deps, creds); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (deps credentialDependencies) get(id CredentialTypeIdentifier, conf *Configuration, creds map[CredentialTypeIdentifier]struct{}) []IssueWizardItem {
+	if _, present := deps[id]; !present {
+		deps[id] = conf.CredentialTypes[id].Dependencies.WizardContents().Choose(conf, creds)
+	}
+	return deps[id]
+}
```

Comment:
If we name it `add`, then the function name does not describe that it not only adds but also returns something. We could use `getAndCache` or something but that seems ugly.

I had hoped that this behaviour would sort of be implicit regardless of the short function name, because if caching it here in the map were not necessary, I would not have created a separate type to attach this method to at all; then I would have used `map[CredentialTypeIdentifier][]IssueWizardItem` directly.

Reply:
To me it was not immediately clear at least. I had to look in detail to the exact code to find out why it worked. But I understand the naming problems. I think a bit different naming would have saved me some time in confusion. A bit of documentation to explain this behaviour would also be sufficient in my opinion. It only is an internal function, so as long as it is clear it is sufficient to me.

Reply:
I have now added explanation in a comment above the function.